### PR TITLE
Bubble in vector invariant form

### PR DESCRIPTION
**Pull request by simonbyrne**

### CODE REVIEW

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
This now runs the bubble for 500 seconds, which blows out the runtime to almost an hour, but it's helpful to see a longer movie.

**Comment by jakebolewski** [comment]
We need to work on being able to optimize through the Curl Operator, that was a remaining TODO from the first pass.

**Comment by jakebolewski** [comment]
We didn't have an issue to track this: #334

**Comment by jakebolewski** [comment]
@szy21 it would be probably more helpful to have a true 3D bubble in this case rather than N slice models for testing the operators (so that there is non-zero mag for each component).

**Comment by szy21** [comment]
> @szy21 it would be probably more helpful to have a true 3D bubble in this case rather than N slice models for testing the operators (so that there is non-zero mag for each vector component).

Sounds good. @akshaysridhar Is there a 3D bubble in the literature that we can follow? Or maybe I could simply add a perturbation in the y direction?

File: examples/hybrid/bubble3d_invariant.jl [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+    @. dw -= vgradc2f(cp) / Ic2f(cρ)
+
+    cE = @. (norm(cuvw)^2) / 2 + Φ(coords.z)
+    @. duₕ -= hgrad(cE)
+    @. dw -= vgradc2f(cE)
+
+    # 3) potential temperature
+    @. dρθ -= vdivf2c(fw * Ic2f(cρθ))
+    @. dρθ -= vdivf2c(Ic2f(cuₕ * cρθ))
+    @. dρθ -= hdiv(cuₕ * Yc.ρθ)
```
Comment (szy21) [inline examples/hybrid/bubble3d_invariant.jl:254]: Is this the same as `@. dρθ -= hdiv(cuₕ * cρθ)`?

File: examples/hybrid/bubble_invariant.jl [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+        top = Operators.Extrapolate(),
+    )
+    cw = If2c.(fw)
+    fuₕ = Ic2f.(cuₕ)
+    cuw = Geometry.Covariant13Vector.(cuₕ) .+ Geometry.Covariant13Vector.(cw)
+
+    dw .= fw .* 0
+
+    # 1.a) horizontal divergence
+    dρ .-= hdiv.(cρ .* (cuw))
```
Comment (szy21) [inline examples/hybrid/bubble_invariant.jl:192]: Why is the horizontal divergence term for ρ different from the one for ρθ (line 249)?

Comment (simonbyrne) [reply examples/hybrid/bubble_invariant.jl:192]: Both work in this case, but this one is more correct: divergence operates on the contravariant components, and in the presence of topography the vertical covariant component may have a horizontal contravariant component.

**Comment by simonbyrne** [comment]
bors try

**Comment by szy21** [comment]
We need to add bubble3d_invariant to ci as well?

**Comment by simonbyrne** [comment]
the plots triggered the out-of-memory!

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

File: examples/hybrid/bubble_2d.jl [hunk 3]
```diff
@@ -327,3 +335,175 @@ Plots.png(
     joinpath(path, "energy.png"),
 )
 Plots.png(Plots.plot((Mass .- mass_0) ./ mass_0), joinpath(path, "mass.png"))
+
```
Comment (szy21) [inline examples/hybrid/bubble_2d.jl:338]: I guess we don't need this (from this line to the end)? Or do we need to change this file at all?

File: examples/hybrid/bubble3d_invariant.jl [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+    )
+    @. dw -= vgradc2f(cp) / Ic2f(cρ)
+
+    cE = @. (norm(cuvw)^2) / 2 + Φ(coords.z)
+    @. duₕ -= hgrad(cE)
+    @. dw -= vgradc2f(cE)
+
+    # 3) potential temperature
+
+    @. dρθ -= hdiv(cuvw * Yc.ρθ)
```
Comment (szy21) [inline examples/hybrid/bubble3d_invariant.jl:253]:
```suggestion
    @. dρθ -= hdiv(cuvw * cρθ)
```

**Pull request merged**